cnap/core/streambroker.py

```diff
@@ -0,0 +1,209 @@
+
+The origin frames from streamprovider were inferenced by inferengine, then will
+be published to stream broker, and then websocket server will fetch frame from
+stream broker for next step action such as displaying on stream dashboard or
+trigger Faas actions.
+
+The `StreamBrokerClientBase` is the abstract class for stream broker client.
+The `RedisStreamBrokerClient` and `KafkaStreamBrokerClient` are subclass for
+Redis and Kafka stream broker server and can be extended to support other types
+of stream broker by implementing other subclasses.
+"""
+
+import logging
+from abc import ABC, abstractmethod
+import time
+
+import redis
+from kafka import KafkaProducer
+import kafka.errors
+import cv2
+
+from core.frame import Frame
+
+# pylint: disable=no-member
+
+LOG = logging.getLogger(__name__)
+
+class StreamBrokerClientBase(ABC):
+
+    """
+    The Abstract class for stream broker client.
+    """
+
+    @abstractmethod
+    def connect(self, host: str, port: int) -> bool:
+        """
+        Connect to broker server.
+
+        Args:
+            host: The host ip of broker server.
+            port: The port of broker server.
+
+        Returns:
+            bool: True if the connection is successful, False otherwise.
+        """
+        raise NotImplementedError("Subclasses should implement connect() method.")
+
+    @abstractmethod
+    def publish_frame(self, topic: str, frame: Frame) -> None:
+        """
+        Publish a frame to a topic.
+
+        Args:
+            topic: The topic name to publish to.
+            frame: The frame to publish.
+
+        Returns: None
+
+        Raises:
+            ValueError: if the topic or frame is None.
+            RuntimeError: if any errors while encoding before publishing frame.
+        """
+        raise NotImplementedError("Subclasses should implement publish_frame() method")
+
+
+class RedisStreamBrokerClient(StreamBrokerClientBase):
+
+    """
+    Redis implementation for stream broker client.
+    """
+
+    def __init__(self):
+        self._conn = None
+
+    def connect(self, host: str="127.0.0.1", port: int=6379):
+        """
+        Connect to Redis server, will attempt to reconnect when a connection error occcurs.
+
+        Args:
+            host: The host hostname/ip of Redis server.
+            port: The port of Redis server.
+
+        Returns: None
+
+        Raises:
+            redis.exceptions.ConnectionError: if connection to the Redis server fails
+                and reconnection exceeds the limit.
+        """
+        self._conn = redis.Redis(host=host, port=port, db=0)
+
+        sleep_time = 0.5
+
+        # Attempts to reconnect when a connection error occurs, up to MAX_RECONNECTION_TIMES times,
+        # so if a connection error still occurs on the (MAX_RECONNECTION_TIMES + 1)th loop,
+        # raise conncection error
+        for i in range(self.MAX_RECONNECTION_TIMES + 1):
```

Comment:
Also update it when you re-set the content value.

Reply:
Declare it in base class.